Polyglot Kotlin with complete Maven model support and other goodies

thorntonrp opened the pull request:
This pull request leverages the previous work that was done and re-worked it to support the complete Maven model as well as the following features:
- Updates Kotlin to 1.3.21
- Full Maven model support (including support for previously unsupported elements)
- Gets the classpath for the Kotlin script engine from the Maven ClassRealm
- Includes support for execute tasks both with inline lambdas and external scripts.
- Unit tests provide 100% coverage for the Maven model.
- Adds integration tests
- Resolves ClassRealm issues that affected IntelliJ IDEA

mosabua commented:
Whoa! This is awesome! Can you get the Takari CLA signed please http://takari.io/support/TakariCLA.pdf

I will give it review in the next days and we can aim to get a new release out shortly after.

mosabua commented:
Can you give this a whirl with your Kotlin usage @Tradunsky please

mosabua commented on .mvn/wrapper/maven-wrapper.properties:
```diff
@@ -1 +1 @@
-distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.5.4/apache-maven-3.5.4-bin.zip
\ No newline at end of file
+distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.3.9/apache-maven-3.3.9-bin.zip
```
please revert .. if anything we want to upgrade to a newer wrapper version.. not go back older..

thorntonrp replied:
It's been reverted to 3.5.4. I pushed it back to 3.3.9 because many IDEs are still baselined on Maven 3.3, but I know IntelliJ at least will be baselining on Maven 3.6.0 in its next release.

mosabua commented:
And for starters.. does this work with the examples project? https://github.com/takari/polyglot-maven-examples

thorntonrp commented:
Contributor agreement has been sent via email.

thorntonrp commented:
As some background on this PR, I'm the technical lead engineer at a large non-profit. My team maintains a build ecosystem that is heavily invested in Maven and includes hundreds of Maven projects. I've evaluated Polyglot Maven recently for my organization as a means to introduce custom scripting into our builds without having to change our build ecosystem to Gradle. My organization is also increasing its investment in Kotlin as one of our primary development languages, so in order to give my team's evaluation the best chance of success, I wanted to make sure that the Kotlin implementation was functionally complete and fully tested. For the last few weeks, I've been actively testing these enhancements against several of my organization's Maven builds and addressing issues as they arise.

thorntonrp commented:
I submitted the PR on the polyglot-maven-examples project:

https://github.com/takari/polyglot-maven-examples/pull/11

That PR updates the Kotlin example to demonstrate the updated DSL and execute features. There were some DSL idioms and overloaded operations from the previous release that I chose not to bring forward because they were either non-idiomatic kotlin, their usage was confusing, or because they didn't appear to add value over their alternatives. That said, I'd be happy to add those back in if they are seen as necessary to preserve backward compatibility or if there is strong argument for their continued inclusion.

Tradunsky commented on polyglot-kotlin/src/test/resources/unit-tests/artifactId/variation-4/pom.kts:
```diff
@@ -0,0 +1,3 @@
+project {
+    id(artifactId = "sample-lib")
```
This looks miss-leading, like id of artifact id. WDYT?

thorntonrp replied:
With just artifactId, it does look a little weird, but it's a unit test for a convenience operation that accepts groupId, artifactId, version, and packaging as separate parameters. The artifactId is the only one that should never be null in this context, so I can set that to required. As it currently stands, the method signature is:
```kotlin
fun id(groupId: String? = null, artifactId: String? = null, version: String? = null, packaging: String? = null) {...}
```

Tradunsky replied:
Yes, I do understand the reasons, I'm asking about DSL design. Not sure if is such thing as id of the project that includes packaging at least. 

Would be this a better design if it will exists in the scope of `project` just like in java DSL?

Tradunsky replied:
Didn't know that other DSLs supports the same approach, ok. Make sense to keep it consistent

Tradunsky commented on polyglot-kotlin/src/test/resources/unit-tests/dependencies/variation-1/pom.kts:
```diff
@@ -0,0 +1,49 @@
+                    groupId = "junit"
+                    artifactId = "junit"
+                }
+                exclusion {
```
This is inconsistent with another style, where it calls "excluding"

thorntonrp replied:
I agree. It just felt weird when invoking it as a function for it not to be a verb.

mosabua replied:
Making it all into verbs is going to be a large sweeping change .. I would rather not do that at this stage.. so backing out and just sticking with exclusion for now seems good.

Tradunsky commented on polyglot-kotlin/src/test/resources/unit-tests/dependencies/variation-8/pom.kts:
```diff
@@ -0,0 +1,17 @@
+project {
+    profiles {
+        profile {
+            id("other")
```
In order to keep the same style with `plugin("gav")` I suggest to have the same here profile("other"). The same was done to `project("id")` already

thorntonrp replied:
This style is actually supported, but it looks like it isn't covered in the unit tests. I'll correct that.

Tradunsky replied:
ok. From my point of view, id as a property looks better comparing to function call

Tradunsky commented on poms/pom.kts:
```diff
@@ -0,0 +1,59 @@
+
+    build {
+        //
+        // Arbitrary Groovy code can be executed in any phase in the form of a dynamic plugin
```
Did you forgotten to change it to Kotlin when pasting from Groovy example?

thorntonrp replied:
Yes. I'll correct that.

Tradunsky commented:
That is really cool that you found time to work on better version of kotlin dsl for polyglot project, I do appreciate it 👍 

Also it is nice that you have production project using it, that is a good push towards pom.kts popularity.


When I started this module I was playing with different pom styles in order to not only map kotlin script to maven model, but to leverage both Intellij Idea and kotlin DSL capabilities to have build files written with more productivity. BTW, I see you have PomDsl annotation 👍 In most of the cases it means that we have compiler taking care of methods available in some function scope {}, which is almost as good as having IDE highlighting on xsd schema :)

One of the style I consider has decreased LOC for builds files significantly (from 215 to 70 lines) and from my point of view with a decent level of readability: https://github.com/takari/polyglot-maven/pull/179/files

mosabua commented on polyglot-kotlin/README.md:
````diff
@@ -1,57 +1,106 @@
+
 ```kotlin
-project {
-    name = "Polyglot :: Kotlin"
````
So is this longer and explicit (and more consistent setup with the original pom) no longer supported? If so why?

thorntonrp replied:
That remains supported. Under `polyglot-kotlin/src/test/resources/unit-tests/name/`, there are three different ways the name can be supplied.

mosabua replied:
Great

mosabua commented on polyglot-kotlin/README.md:
````diff
@@ -1,57 +1,106 @@
+
 ```kotlin
-project {
-    name = "Polyglot :: Kotlin"
````
So is this longer and explicit (and more consistent setup with the original pom) no longer supported? If so why?

thorntonrp replied:
There are 3 different ways to supply the project name. That one is still supported. See `polyglot-kotlin/src/test/resources/unit-tests/name` for each supported variation.

thorntonrp replied:
Variation 1:
```kotlin
project {
    name = "Sample Library"
}
```
Variation 2
```kotlin
project("Sample Library") {
} 
```
Variation 3
```kotlin
project {
    name("Sample Library")
}
```

thorntonrp replied:
The unit tests provide examples of each variation for each model element.

mosabua replied:
Great.. thanks for confirming.

mosabua commented on polyglot-kotlin/src/main/kotlin/org/sonatype/maven/polyglot/kotlin/serialization/ModelScriptWriter.kt:
````diff
@@ -0,0 +1,916 @@
+ * fields. All complex model elements can be configured using this "block" flavor of the DSL. The "block" flavor,
+ * however, does little on its own to reduce the code verbosity of the original XML.
+ *
+ * ```
````
Okay ... I understand better now how you have two .. one close to maven original pom, one closer to kotlin .. question regarding that .. can they be mixed to some degree or do you have to choose? If I understand correctly you cant mix ..

thorntonrp replied:
The flavors can be mixed with no affect on functionality. Kotlin is able to infer the correct function to be called.

mosabua commented on polyglot-kotlin/src/main/kotlin/org/sonatype/maven/polyglot/kotlin/serialization/ModelScriptWriter.kt:
````diff
@@ -0,0 +1,916 @@
+ *
+ * ```
+ * project("Sample Application") {
+ *   id("com.example:sample-app:1.0:jar")
````
using gav or so might be an alternative

thorntonrp replied:
I chose "id" as the function name to be consistent with some of the other polyglot-maven implementations (atom, ruby, etc) that also provide a field or function named "id" for the same purpose. I'm not opposed to renaming it, but perhaps there needs to be a broader discussion around common features that should be consistent between implementations.

mosabua replied:
fair enough .. lets leave it as is for now then. The large discussion might have to be done at some stage.. or we just stick with this .. it works as well.

mosabua commented on polyglot-kotlin/src/main/kotlin/org/sonatype/maven/polyglot/kotlin/serialization/ModelScriptWriter.kt:
````diff
@@ -0,0 +1,916 @@
+ * order to minimize the verbosity of specific model elements.
+ *
+ * ```
+ * project("Sample Application") {
````
is name required here? because typically in Maven it is not .... maybe the parameter in that constructor should be the typical required ones of GAV

thorntonrp replied:
No argument is required on the project. When it is supplied, it can be either the "name" or the GAV.

mosabua commented:
I gave some initial feedback... you seem to have done a lot of reformatting and such in other dialect and unrelated changes.. this makes this PR unnecessarily hard and huge to review. Could you maybe back a lot of those unrelated changes out? Otherwise I will have to spend quite a bit more time reviewing before I can merge this .

thorntonrp commented:
Okay

On Sat, Mar 2, 2019, 6:37 PM Manfred Moser <notifications@github.com> wrote:

> I gave some initial feedback... you seem to have done a lot of
> reformatting and such in other dialect and unrelated changes.. this makes
> this PR unnecessarily hard and huge to review. Could you maybe back a lot
> of those unrelated changes out? Otherwise I will have to spend quite a bit
> more time reviewing before I can merge this .
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/takari/polyglot-maven/pull/182#issuecomment-468977663>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD4e5wOZpDhpZJM_fmcoojFfXvh5WFFRks5vSyd1gaJpZM4bTizP>
> .
>

thorntonrp commented:
@Tradunsky My purpose in inheriting from the Java Maven model instead of re-creating it in Kotlin was to ensure full support of (and compatibility with) the Java Maven model. Now that we have 100% test coverage of the model to catch any regressions in core functionality, I would not be opposed to refactoring the model to no longer inherit from the Java model.

mosabua commented:
Thanks for backing out the unrelated changes.. I am more than happy to do another PR or whatever later that is purely a reformat of pom files or whatever ..

mosabua commented:
Given that this is really a high quality but large PR I am inclined to merge this as is. Then we can do refinements where needed with new PRs and maybe release in a week or so? What do think?

Tradunsky commented:
@thorntonrp Regarding inheritance - it is up to you. Just sharing my thoughts around DSL usability and auto-completion in IDE. However, even if new features added to original model, they would not be necessary kotlin idiomic, so not sure how it helps in supporting a new changes, since you still need to make their support in 3 different DSL approaches that chosen.

I know this DSLs not as commonly used as xml, but when you define some API, like calling original method or some variation of specifying project id, you have to keep supporting it or make a smooth migration for users, so they do not punished for choosing kotlin dsl instead of standard xml.

Overall work looks good to me, there are decent amount of tests and features. I think what @mosabua suggested does make sense.

thorntonrp commented:
@mosabua Is there anything else you need from me in order to complete this pull request?
